### PR TITLE
DROOLS-4061 : Guided Score card rules not executed via test scenario.

### PR DESCRIPTION
**Pull request by Rikkola**
https://issues.jboss.org/browse/DROOLS-4061

https://github.com/kiegroup/drools/pull/2356
https://github.com/kiegroup/kie-wb-common/pull/2657
https://github.com/kiegroup/drools-wb/pull/1147

### CODE REVIEW

**Comment by Rikkola** [comment]
@jomarko and others, this PR indexes the scorecard model name and adds find for all scorecards in a specified package.

**Comment by jomarko** [comment]
@Rikkola I will proceed with the review, could you please have a look on https://github.com/Rikkola/kie-wb-common/pull/5 in the meantime?

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                sharedPartIndexTerm = (ValueSharedPartIndexTerm) term;
+            } else if (term instanceof ValueModuleRootPathIndexTerm) {
+                moduleRootPathIndexTerm = (ValueModuleRootPathIndexTerm) term;
+            } else if (term instanceof ValuePackageNameIndexTerm) {
+                packageNameIndexTerm = (ValuePackageNameIndexTerm) term;
+            }
+        }
+
+        final SearchEmptyQueryBuilder queryBuilder = new SearchEmptyQueryBuilder()
+                .addTerm(sharedPartIndexTerm);
```
Comment (jomarko) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: Null check also here?
```java
if (sharedPartIndexTerm != null) {
            queryBuilder.addTerm(sharedPartIndexTerm);
}
```

Comment (jomarko) [reply kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: Second proposal:
```java
    @Override
    public Query toQuery(Set<ValueIndexTerm> terms) {
        final SearchEmptyQueryBuilder queryBuilder = new SearchEmptyQueryBuilder();

        for (ValueIndexTerm term : terms) {
            if (term instanceof ValueSharedPartIndexTerm) {
                queryBuilder.addTerm(term);
            } else if (term instanceof ValueModuleRootPathIndexTerm) {
                queryBuilder.addTerm(term);
            } else if (term instanceof ValuePackageNameIndexTerm) {
                queryBuilder.addTerm(term);
            }
        }

        return queryBuilder.build();
    }
```

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+        checkInvalidAndRequiredTerms(queryTerms,
+                                     NAME,
+                                     new String[]{
+                                             null, null,  // not required
+                                             scorecardTerm.getTerm()
+                                     },
+                                     (t) -> (t instanceof ValueModuleRootPathIndexTerm),
+                                     (t) -> (t instanceof ValuePackageNameIndexTerm),
+                                     (t) -> (t.getTerm().equals(scorecardTerm.getTerm())));
```
Comment (jomarko) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: Please `Objects.equals(t.getTerm(), scorecardTerm.getTerm())`

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                                                              final int startRow,
+                                                              final List<KObject> kObjects) {
+            final int hits = kObjects.size();
+            final PageResponse<RefactoringPageRow> response = new PageResponse();
+            final List<RefactoringPageRow> result = buildResponse(kObjects);
+            response.setTotalRowSize(hits);
+            response.setPageRowList(result);
+            response.setTotalRowSizeExact(true);
+            response.setStartRowIndex(startRow);
+            response.setLastPage((pageSize * startRow + 2) >= hits);
```
Comment (jomarko) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: I think I do not understand this, for example,
```java
pageSize = 10;
startRow = 10;
hits = 23;

// 102 >= 23 - but I think is second page of three
```

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        }
+
+        private Set<String> getScorecardModelNamesNamesFromKObject(final KObject kObject) {
+            Set<String> result = new HashSet<>();
+            if (kObject == null) {
+                return result;
+            }
+            for (KProperty property : kObject.getProperties()) {
+
+                if (Objects.equals("shared:scorecardModelName", property.getName())) {
```
Comment (jomarko) [inline kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: :heart:

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java (hunk 3 above)
Comment (Rikkola) [reply kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: Thinking and I think I need to drop the paging from here and just do one big list. That of course does not work for lists with 100 scorecards, but if you have a setup like that then I guess that UI issue is not the only one you have.

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java (hunk 4 above)
Comment (Rikkola) [reply kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: Slowly learning ;)

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java (hunk 3 above)
Comment (Rikkola) [reply kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: I don't know, if this is an issue then we have the same problem in every page query :)

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
Jenkins please test full downstream

**Comment by cristianonicolai** [comment]
jenkins please retest this

**Comment by Rikkola** [comment]
Migration tool failing.
jenkins please retest this

**Comment by Rikkola** [comment]
Jenkins please test full downstream

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-api/src/main/java/org/kie/workbench/common/services/refactoring/service/PartType.java [hunk 5]
```diff
@@ -120,6 +120,15 @@
      */
     FORM_FIELD("formfield"),
 
+    /*
+     * Scorecards
+     */
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-api/src/main/java/org/kie/workbench/common/services/refactoring/service/PartType.java]: No the file has a weird grouping of comments. First a comment like this for what the fields belong to and then comments for the actual fields.

File: kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        final SearchEmptyQueryBuilder queryBuilder = new SearchEmptyQueryBuilder();
+
+        for (ValueIndexTerm term : terms) {
+            if (term instanceof ValueSharedPartIndexTerm) {
+                queryBuilder.addTerm(term);
+            } else if (term instanceof ValueModuleRootPathIndexTerm) {
+                queryBuilder.addTerm(term);
+            } else if (term instanceof ValuePackageNameIndexTerm) {
+                queryBuilder.addTerm(term);
+            }
```
[the remark this answers is not in the document]
Comment (Rikkola) [reply kie-wb-common-services/kie-wb-common-refactoring/kie-wb-common-refactoring-backend/src/main/java/org/kie/workbench/common/services/refactoring/backend/server/query/standard/FindScorecardNamesQuery.java]: The validateTerms method below should make sure there are only supported terms in the list.

**Comment by Rikkola** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
jenkins execute full downstream build

org.kie.wb.selenium.ui.LoadAllPerspectivesIntegrationTest.allPerspectivesCanBeLoaded
org.kie.wb.selenium.ui.LoginIntegrationTest.loginAndLogout
org.kie.wb.selenium.ui.ProjectLibraryIntegrationTest.importAndBuildProjectFromCustomRepository

**Comment by Rikkola** [comment]
jenkins execute full downstream build

Same tests failing.

**Comment by Rikkola** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
Not sure why I even bother, but here we go again.

jenkins execute full downstream build

**Comment by Rikkola** [comment]
The usual Selenium tests failed.

jenkins execute full downstream build

**Comment by Rikkola** [comment]
I don't think I have had a full downstream build that passed the whole week and I'm firing several each day :)

"org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.codehaus.mojo:gwt-maven-plugin:2.8.2:compile (gwt-compile) on project jbpm-wb-case-mgmt-showcase: Command failed with status 1"

jenkins execute full downstream build

**Comment by manstis** [comment]
@Rikkola Perhaps @nmirasch or @cristianonicolai need to be made aware!?

**Comment by Rikkola** [comment]
Perhaps if this continues. I can't say I trust any error that shows up for the first time. Second or third time and it is worth looking into.

**Comment by Rikkola** [comment]
jenkins execute full downstream build

`12:11:54 org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M1:enforce (ban-blacklisted-dependencies) on project kie-wb-common-compiler-distribution: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed.`

**Comment by Rikkola** [comment]
13 failures in a row

**Comment by Rikkola** [comment]
drools-wb needed a rebase.

jenkins execute full downstream build

**Comment by Rikkola** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers Can I get a merge please. Haven't had a green full downstream build during my last 20-30 builds, so I doubt it will ever happen here. Everything else is green and the existing issue with the full build is known to fail and is not related to these PRs.

**Comment by Rikkola** [comment]
jenkins execute full downstream build

org.kie.wb.selenium.ui.ProjectLibraryIntegrationTest.importAndBuildProjectFromStockRepository
org.kie.wb.selenium.ui.StandalonePerspectivesIntegrationTest.testStandaloneEditorPerspective

**Comment by danielezonca** [comment]
jenkins execute full downstream build

Same two failures but not they should (could) be green because of https://github.com/kiegroup/kie-wb-distributions/pull/952
org.kie.wb.selenium.ui.ProjectLibraryIntegrationTest.importAndBuildProjectFromStockRepository
org.kie.wb.selenium.ui.StandalonePerspectivesIntegrationTest.testStandaloneEditorPerspective

**Comment by jomarko** [comment]
Was the last build here started before or after kiegroup/kie-wb-distributions#952 merge?

**Comment by danielezonca** [comment]
@jomarko I'm quite sure before, that's why I asked now for another build

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by romartin** [comment]
Hey @Rikkola plz lemme know once builds are fine and I'll merge, np! Thanks man!!

**Comment by jomarko** [comment]
jenkins execute full downstream build

**Comment by Rikkola** [comment]
@jomarko probably not worth it to try and get this green. Haven't seen a green full build for months. (but I hear there was one last week). I'm just trying to get different random failures to prove this change did not break anything.

**Comment by jomarko** [comment]
@Rikkola again, just selenium tests failed. In the meantime https://github.com/kiegroup/kie-wb-distributions/pull/942 was merged, last try of full build here

jenkins execute full downstream build

**Comment by Rikkola** [comment]
jenkins please retest this

"org.kie.workbench.screens.workbench.backend.BaseAppSetupTest.setupUndefinedConfigurationGroupTest" failed
This is new, but not sure how it could be caused by me.

**Comment by jomarko** [comment]
@Rikkola hi, regarding `BaseAppSetupTest`, I investigate it today, here is the summary,
- It started fail on jenkins `dailyBuild` job on 9 of August
- It fails on place where `fileSystem` produced by `FileSystemTestingUtils` is expected, but is null
- Locally I am not able to reproduce
- I am trying to understand, while `fileSystem` is null by this PR https://github.com/kiegroup/appformer/pull/768, where I removed try catch block and hope there will be an exception telling us details

**Comment by Rikkola** [comment]
Full build failed with "Caused by: org.apache.maven.plugin.compiler.CompilationFailureException: Compilation failure
/home/jenkins/workspace/KIE/master/pullrequest/kie-wb-common-downstream-pullrequests/kie-wb-common-screens/kie-wb-common-default-editor/kie-wb-common-default-editor-client/src/main/java/org/kie/workbench/common/screens/defaulteditor/client/editor/NewFileUploader.java:[132,28] incompatible types: incompatible parameter types in lambda expression" 
So out of sync.

@romartin @kiegroup/gatekeepers I think it is safe to merge these in. The changes are already in 7.25.x

**Comment by Rikkola** [comment]
jenkins please retest this

**Comment by Rikkola** [comment]
"org.kie.workbench.screens.workbench.backend.BaseAppSetupTest.setupUndefinedConfigurationGroupTest"

**Comment by mareknovotny** [comment]
@Rikkola this PR https://github.com/kiegroup/kie-wb-common/pull/2821 should fix that failing test

**Comment by Rikkola** [comment]
jenkins please retest this

**Comment by Rikkola** [comment]
Jenkins please test full downstream

**Comment by Rikkola** [comment]
"org.kie.hacep.RemoteStreamingKieSessionImplTest.getListKieSessionObjectsTest" 
From the full downstream.

**Comment by Rikkola** [comment]
@kiegroup/gatekeepers Can I get a merge please. The full build refuses to pass, but the errors are not related to these changes and keep changing. So this PR bundle should be ok.

**Comment by romartin** [comment]
Jenkins execute full downstream build

**Comment by Rikkola** [comment]
@romartin Your a magician! Green full downstream build. I started 18 of them for this PR and another 10 or so for other builds and have not seen a green one for few months!

**Pull request merged**